presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxBy.java

```diff
@@ -95,40 +144,139 @@ private InternalAggregationFunction generateAggregation(Type valueType, Type key
 
     private static List<ParameterMetadata> createInputParameterMetadata(Type value, Type key)
     {
-        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, key), new ParameterMetadata(BLOCK_INDEX));
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INPUT_CHANNEL, key), new ParameterMetadata(BLOCK_INDEX));
     }
 
-    public static void input(boolean min, Type keyType, MaxOrMinByState state, Block value, Block key, int position)
-    {
-        if ((state.getKey() == null) || state.getKey().isNull(0) ||
-                compare(keyType.compareTo(key, position, state.getKey(), 0), min)) {
-            state.setKey(key.getSingleValueBlock(position));
-            state.setValue(value.getSingleValueBlock(position));
+    private StateCompiler.StateClass generateStateClass(Type keyType, Type valueType, DynamicClassLoader classLoader) {
```

Comment:
Is there a specific reason to generate these interfaces programmatically? Assuming the number of interfaces is bounded (and a function of the native types Presto supports), this is way too complicated for what we get out of it. It makes the code more brittle, harder to follow, debug, etc.